cmd/infracost/run.go

```diff
@@ -802,14 +803,116 @@ func AddHCLEnvVars(projectContexts []*config.ProjectContext, r output.Root, hclR
 	env["tfRunTimeMs"] = tfTimeTaken
 	env["hclRunTimeMs"] = hclTimeTaken
 
+	env["hclMissingResources"] = hclMissingResources(projects, hclProjects)
+	env["hclResourceDiff"] = hclResourceDiff(projects, hclProjects)
+}
+
+func hclMissingResources(projects, hclProjects []*schema.Project) []string {
+	missingResources := map[string]bool{}
+
+	hclProjectsMapping := map[string]*schema.Project{}
+	for _, project := range hclProjects {
+		hclProjectsMapping[project.Name] = project
+	}
+
+	for _, project := range projects {
+		hclProject := hclProjectsMapping[project.Name]
+
+		if hclProject == nil {
+			log.Debugf("could not find a matching HCL project for '%s' on missing resources search", project.Name)
+			continue
+		}
+
+		hclResourceNames := map[string]bool{}
+		for _, resource := range hclProject.Resources {
+			hclResourceNames[resource.Name] = true
+		}
+
+		for _, resource := range project.Resources {
+			if _, ok := hclResourceNames[resource.Name]; !ok {
+				missingResources[resource.ResourceType] = true
+			}
+		}
+	}
+
+	missingList := make([]string, 0, len(missingResources))
+	for key := range missingResources {
+		missingList = append(missingList, key)
+	}
+
+	return missingList
+}
+
+func hclResourceDiff(projects, hclProjects []*schema.Project) map[string][]string {
+	diff := map[string][]string{}
+
+	hclProjectsMapping := map[string]*schema.Project{}
+	for _, project := range hclProjects {
+		hclProjectsMapping[project.Name] = project
+	}
+
+	for _, project := range projects {
+		hclProject := hclProjectsMapping[project.Name]
+
+		if hclProject == nil {
+			log.Debugf("could not find a matching HCL project for '%s' on resources diff", project.Name)
+			continue
+		}
+
+		acc := map[string]*decimal.Decimal{}
+
+		for _, resource := range project.Resources {
+			acc[resource.Name] = resource.MonthlyCost
+		}
+
+		for _, hclResource := range hclProject.Resources {
+			resourceCost, ok := acc[hclResource.Name]
+			if !ok {
+				continue
+			}
+
```

Comment:
I would prefer here if we merged the logic from `hclMissingResources` so that we don't have create lookup maps and loops twice. i.e. here we could just add to the `missingList`. 

We can alter the return signature of the `hclResourceDiff` to return a struct with missing resources and discrepancies, e.g:

```go
type hclProjectDiff struct {
    resourceDiffs map[string][]string
    missingResources []string
}
```

Reply:
Sounds good, I'll name it `hclRunDiff` though as it collects data from all projects. Sounds good?

Reply:
👍